lib/sdam/topology.js

```diff
@@ -0,0 +1,416 @@
+'use strict';
+const EventEmitter = require('events');
+const ServerDescription = require('./server_description').ServerDescription;
+const TopologyDescription = require('./topology_description').TopologyDescription;
+const TopologyType = require('./topology_description').TopologyType;
+const monitoring = require('./monitoring');
+const calculateDurationInMs = require('../utils').calculateDurationInMs;
+const MongoTimeoutError = require('../error').MongoTimeoutError;
+
+// Global state
+let globalTopologyCounter = 0;
+
+// Constants
+const DEFAULT_LOCAL_THRESHOLD_MS = 15;
+const DEFAULT_HEARTBEAT_FREQUENCY = 10000;
+const DEFAULT_SERVER_SELECTION_TIMEOUT = 30000;
+
+/**
+ * A container of server instances representing a connection to a MongoDB topology.
+ *
+ * @fires Topology#serverOpening
+ * @fires Topology#serverClosed
+ * @fires Topology#serverDescriptionChanged
+ * @fires Topology#topologyOpening
+ * @fires Topology#topologyClosed
+ * @fires Topology#topologyDescriptionChanged
+ * @fires Topology#serverHeartbeatStarted
+ * @fires Topology#serverHeartbeatSucceeded
+ * @fires Topology#serverHeartbeatFailed
+ */
+class Topology extends EventEmitter {
+  /**
+   * Create a topology
+   *
+   * @param {Array|String} seedlist a string list, or array of Server instances to connect to
+   * @param {Object} [options] Optional settings
+   * @param {Number} [options.localThresholdMS=15] The size of the latency window for selecting among multiple suitable servers
+   * @param {Number} [options.serverSelectionTimeoutMS=30000] How long to block for server selection before throwing an error
+   * @param {Number} [options.heartbeatFrequencyMS=10000] The frequency with which topology updates are scheduled
+   */
+  constructor(seedlist, options) {
+    super();
+    seedlist = seedlist || [];
+    options = Object.assign(
+      {},
+      {
+        localThresholdMS: DEFAULT_LOCAL_THRESHOLD_MS,
+        serverSelectionTimeoutMS: DEFAULT_SERVER_SELECTION_TIMEOUT,
+        heartbeatFrequencyMS: DEFAULT_HEARTBEAT_FREQUENCY
+      },
+      options
+    );
+
+    const topologyType =
+      seedlist.length === 1 && !options.replicaset
+        ? TopologyType.Single
+        : options.replicaset
+          ? TopologyType.ReplicaSetNoPrimary
+          : TopologyType.Unknown;
+
+    const topologyId = globalTopologyCounter++;
```

Comment:
Can we make a `topologyCounter()`?

Reply:
what do you mean by this? a function on the class itself? a package-local variable seems fine to me, what would we gain otherwise?

Reply:
I just don't like the idea of incrementing the `globalTopologyCounter` inline. I'd rather that we wrap it in a `topologyCounter` function that internally handles it.